### ISSUE TITLE
How to use this library with a custom clock?

### ISSUE THREAD

**Issue by alkatar21** [HowardHinnant/date#835]
I have a custom clock and I have problems getting it to work with this library.  As far as I understand it currently, all relevant functions are overloaded for `sys_time` and `local_time`. How do I use the library for own clocks? Do I have to overload all functions myself, or is there a smarter way to replace a system clock with my own clock?

**Comment by HowardHinnant** [comment]
Here is a custom clock with milliseconds precision, based on `system_clock`, but counting time since 2000-01-01:

```c++
#include "date/tz.h"
#include <chrono>
#include <cstdint>
#include <ostream>

struct my_clock;

template <class Duration>
    using my_time = std::chrono::time_point<my_clock, Duration>;

struct my_clock
{
    using duration   = std::chrono::milliseconds;
    using rep        = duration::rep;
    using period     = duration::period;
    using time_point = std::chrono::time_point<my_clock>;

    static constexpr bool is_steady = false;

    static time_point now() noexcept;

    template <class Duration>
    static
    auto
    from_sys(date::sys_time<Duration> const& tp) noexcept;

    template <class Duration>
    static
    auto
    to_sys(my_time<Duration> const& tp) noexcept;
};

template <class Duration>
auto
my_clock::from_sys(date::sys_time<Duration> const& tp) noexcept
{
    using namespace date;
    auto constexpr epoch = sys_days{January/1/2000};
    return my_time<Duration>{tp - epoch};
}

template <class Duration>
auto
my_clock::to_sys(my_time<Duration> const& tp) noexcept
{
    using namespace date;
    return sys_time<Duration>{tp - clock_cast<my_clock>(sys_days{})};
}

my_clock::time_point
my_clock::now() noexcept
{
    using namespace date;
    using namespace std::chrono;
    return clock_cast<my_clock>(floor<my_clock::duration>(system_clock::now()));
}

template <class Duration>
std::ostream&
operator<<(std::ostream& os, my_time<Duration> const& tp)
{
    using namespace date;
    using namespace std::chrono;
    return os << clock_cast<system_clock>(my_clock::now());
}

#include <iostream>

int
main()
{
    std::cout << my_clock::now() << '\n';
}
```

`my_clock` introduces the nested types that indicate it is coining `std::chrono::milliseconds`.  It then implements the static member functions `from_sys`, `to_sys` and `now`.

The epoch, 2000-01-01, need only be mentioned in `from_sys`.  This function converts a `sys_time` to a `my_time`.  It does this by subtracting the input `sys_time` from 2000-01-01 and casting that `duration` to a `my_time`.

`to_sys` reuses the epoch in `from_sys` by taking the `sys_time` epoch (`sys_days{}`), converting that to `my_time` (`clock_cast<my_clock>` will call `my_clock::from_sys`), subtracting that result from the input `my_time`, and converting that `duration` to `sys_time`.

`now()` simply gets the current current time using `system_clock`, `floor`s it to the proper precision, and converts it into `my_time`.

A convenience `operator<<` is set up to print out the time_point by converting it to `sys_time` and use `sys_time`'s streaming operator.

**Comment by alkatar21** [comment]
Thanks for the detailed answer. I had picked up a lot of this from other posts, but it's good to have it condensed. Do I understand correctly that the suggestion here is to convert to `system_clock` each time and to use this lib (`data::format`, `date::year_month_day { tp }`, etc.)?

However, I don't think I gave enough context in the original question for my question to make sense. I have to use C++17 and therefore use this library instead of `std::chrono`. `my_clock::now()` is supposed to return either `system_clock::now()` or under certain conditions times from the past/future (are actually also `system_clock::time_point` instances). I had understood that I have to define my own clock for this, even if I only have to modify what time_point now (I do not want to change the epoch). This time is then used to calculate or format with this library. But I can't use the functions with my own clock and I would like to avoid converting manually every time.

So is it intended to convert to `sys_time` every time, or is there a more elegant solution?
